### PR TITLE
[FLINK-10932] [ResourceManager] Initial flink-kubernetes module with empty implementation

### PR DESCRIPTION
**Pull request by isunjin**
## What is the purpose of the change

Initialize the skeleton module to start native k8s integration

## Brief change log

  - *Add flink-kubernetes module*
  - *Add first Implement of ClusterDescriptor*
  - *Add Interface of Kubernetes client*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no )
  - The runtime per-record code paths (performance sensitive): ( no )
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no )
  - The S3 file system connector: ( no )

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (the document will be introduced in other pull request)

### CODE REVIEW

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cluster/KubernetesClusterDescriptor.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+	private FlinkKubernetesOptions options;
+
+	private KubeClient client;
+
+	public KubernetesClusterDescriptor(@Nonnull FlinkKubernetesOptions options, @Nonnull KubeClient client) {
+		this.options = options;
+		this.client = client;
+	}
+
+	private String generateClusterId(){
```
Comment (TisonKun) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cluster/KubernetesClusterDescriptor.java]: `) {`

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cluster/KubernetesClusterDescriptor.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+			this.client.logException(e);
+			this.tryKillCluster(clusterId);
+			throw new ClusterDeploymentException("Could not create Kubernetes cluster " + clusterId, e);
+		}
+	}
+
+	/**
+	 * Try to kill cluster without throw exception.
+	 * */
+	private void tryKillCluster(String clusterId){
```
Comment (TisonKun) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cluster/KubernetesClusterDescriptor.java]: `) {`

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cluster/KubernetesClusterDescriptor.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+		try {
+			this.killCluster(clusterId);
+		} catch (Exception e){
```
Comment (TisonKun) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cluster/KubernetesClusterDescriptor.java]: `) {`

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cluster/KubernetesClusterDescriptor.java (hunk 1 above)
Comment (isunjin) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cluster/KubernetesClusterDescriptor.java]: Thanks TisonKun.
Quite amount of code needed to implement this feature, as a result I would like to split the code into small pieces to make review easier.

**Comment by wangyang0918** [comment]
@isunjin It’s great to start the kubernetes integration with flink. I am glad to get involved in and could help to give some suggestions based on our use case.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cluster/KubernetesClusterDescriptor.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+		}
+	}
+
+	@Override
+	public void killCluster(String clusterId) throws FlinkException {
+		try {
+			this.client.stopAndCleanupCluster(clusterId);
+		} catch (Exception e) {
+			this.client.logException(e);
+			throw new FlinkException("Could not create Kubernetes cluster " + clusterId);
```
Comment (f7753) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cluster/KubernetesClusterDescriptor.java]: Is this a typo for "Could not create Kubernetes cluster" in `killCluster` method.

**Comment by TisonKun** [comment]
implemented at 3d94ba0fae6a7391c03b7eddb087a68bf178b45d

**Pull request closed without merging**